### PR TITLE
[shadow-api] Add experimental shadow api MAPSAND-165

### PR DESCRIPTION
**Pull request by ank27**
### Summary of changes
Added experimental shadow apis. 

### User impact (optional)

Screenshot 2022-06-22 at 14 06 29

## Pull request checklist:
 - [x] Briefly describe the changes in this PR.
 - [ ] Include before/after visuals or gifs if this PR includes visual changes.
    
 - [ ] Write tests for all new functionality. If tests were not written, please explain why.
 - [ ] Optimize code for java consumption (`@JvmOverloads`, `@file:JvmName`, etc).
 - [x] Add example if relevant.
 - [ ] Document any changes to public APIs.
 - [ ] Run `make update-api` to update generated api files, if there's public API changes, otherwise the `verify-api-*` CI steps might fail.
 - [x] Update [CHANGELOG.md](../CHANGELOG.md) or use the label 'skip changelog', otherwise `check changelog` CI step will fail.
 - [x] If this PR is a `v10.[version]` release branch fix / enhancement, merge it to `main` firstly and then port to `v10.[version]` release branch.

Fixes: < Link to related issues that will be fixed by this pull request, if they exist >

PRs must be submitted under the terms of our Contributor License Agreement [CLA](https://github.com/mapbox/mapbox-maps-android/blob/main/CONTRIBUTING.md#contributor-license-agreement).

### CODE REVIEW

File: extension-style/src/main/java/com/mapbox/maps/extension/style/light/generated/Light.kt [hunk 1]
```diff
@@ -87,6 +88,60 @@ class Light : LightDslReceiver, StyleContract.StyleLightExtension {
...
+     */
+    get() {
+      return getPropertyValue("cast-shadows")
+    }
+  /**
+   * Enable/Disable shadow casting for this light
+   *
+   * @param castShadows as Boolean
+   */
+  override fun castShadows(castShadows: Boolean) = apply {
```
Comment (pengdev) [inline extension-style/src/main/java/com/mapbox/maps/extension/style/light/generated/Light.kt:109]: need mapboxexperimental here as well

Comment (pengdev) [reply extension-style/src/main/java/com/mapbox/maps/extension/style/light/generated/Light.kt:110]: Fixed

File: app/src/main/java/com/mapbox/maps/testapp/examples/terrain3D/Terrain3DShowcaseActivity.kt [hunk 2]
```diff
@@ -40,6 +41,10 @@ class Terrain3DShowcaseActivity : AppCompatActivity() {
         }
         +atmosphere { }
         +projection(ProjectionName.GLOBE)
+        +light {
+          castShadows(true)
+          shadowIntensity(10.0)
+        }
```
Comment (pengdev) [inline app/src/main/java/com/mapbox/maps/testapp/examples/terrain3D/Terrain3DShowcaseActivity.kt:47]: not sure we should add shadow to the terrain activity, and seems the terrain doesn't work well with shadows?

**Review by kiryldz — COMMENTED** [review]
changelog missing and better rebase on main as well

File: app/src/main/java/com/mapbox/maps/testapp/examples/terrain3D/Terrain3DShowcaseActivity.kt (hunk 2 above)
Comment (ank27) [reply app/src/main/java/com/mapbox/maps/testapp/examples/terrain3D/Terrain3DShowcaseActivity.kt]: removed from terrain activity.

**Review by kiryldz — APPROVED** [review]
Instrument tests fail for Shadow API (getters are returning null) - is this expected? We should ticket it out for gl-native then, apart from that code looks OK.

**Comment by ank27** [comment]
> 

yeah, it could be since gl-native hasn't merged and released shadow apis ? 🤔

**Comment by kiryldz** [comment]
> > Instrument tests fail for Shadow API (getters are returning null) - is this expected? We should ticket it out for gl-native then, apart from that code looks OK.
> 
> yeah, it could be since gl-native hasn't merged and released shadow apis ? 🤔

Then those should pass if you run locally, can you confirm that?

**Review by pengdev — APPROVED** [review]

**Pull request merged**